xapi/xapi.go

```diff
@@ -130,66 +130,93 @@ type partialResult struct {
 	aggregate bool
 	fields    []*types.FieldType
 	reader    io.ReadCloser
-	resp      *tipb.SelectResponse
+	rows      []selectResponseRow
 	cursor    int
 
 	done    chan error
 	fetched bool
 }
 
+type selectResponseRow struct {
+	handle int64
+	data   []types.Datum
+}
+
 func (pr *partialResult) fetch() {
-	pr.resp = new(tipb.SelectResponse)
+	resp := new(tipb.SelectResponse)
 	b, err := ioutil.ReadAll(pr.reader)
 	pr.reader.Close()
 	if err != nil {
 		pr.done <- errors.Trace(err)
 		return
 	}
-	err = pr.resp.Unmarshal(b)
+	err = resp.Unmarshal(b)
 	if err != nil {
 		pr.done <- errors.Trace(err)
 		return
 	}
-	if pr.resp.Error != nil {
-		pr.done <- errInvalidResp.Gen("[%d %s]", pr.resp.Error.GetCode(), pr.resp.Error.GetMsg())
+	if resp.Error != nil {
+		pr.done <- errInvalidResp.Gen("[%d %s]", resp.Error.GetCode(), resp.Error.GetMsg())
+		return
 	}
-	pr.done <- nil
-}
 
-// Next returns the next row of the sub result.
-// If no more row to return, data would be nil.
-func (pr *partialResult) Next() (handle int64, data []types.Datum, err error) {
-	if !pr.fetched {
-		select {
-		case err = <-pr.done:
-		}
-		pr.fetched = true
+	pr.rows = make([]selectResponseRow, len(resp.Rows))
```

Comment:
This will call runtime.makeslice(). It cost a lot of time when there are many rows. And it also bring a lot of work for gc.